[WGSL] Make callee ordering deterministic

tadeuzagallo opened the pull request:
#### 0bdc0ccabceb139e0f8fe7498eee77bf63969208
<pre>
[WGSL] Make callee ordering deterministic
<a href="https://bugs.webkit.org/show_bug.cgi?id=257714">https://bugs.webkit.org/show_bug.cgi?id=257714</a>
rdar://110263964

Reviewed by Myles C. Maxfield.

The call graph&apos;s callee list was built from a HashMap, which meant we visited
callees in a non-deterministic order. This makes it harder to write tests that
rely on the order of the output, and we also had crashes that were happening
only when we visited the callees in a certain order. To fix this, we consistently
put callees in a list as we discover them and use the previous hash map to keep
track of where a given callee is on the list so we can add call sites to it later.

* Source/WebGPU/WGSL/CallGraph.cpp:
(WGSL::CallGraphBuilder::visit):

Canonical link: <a href="https://commits.webkit.org/264932@main">https://commits.webkit.org/264932@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6b7d56a9c94bfed92a4f2709f5264bb5134c058a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/9178 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/9459 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/9687 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/10834 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/9100 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/11447 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/9426 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/11958 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/9325 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/10273 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/7970 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/10990 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/7568 "Passed tests") | | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/15838 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/8671 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/8522 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/11839 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/9023 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/7360 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/8252 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2219 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/12475 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/8784 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

litherum commented on Source/WebGPU/WGSL/CallGraph.cpp:
```diff
@@ -98,11 +99,10 @@ void CallGraphBuilder::visit(AST::Function& function)
     if (!result.isNewEntry)
         return;
 
+    m_callees = &result.iterator->value;
```
Should we assert it's null here?

tadeuzagallo replied:
Sounds good

tadeuzagallo committed:
[WGSL] Make callee ordering deterministic
https://bugs.webkit.org/show_bug.cgi?id=257714
rdar://110263964

Reviewed by Myles C. Maxfield.

The call graph's callee list was built from a HashMap, which meant we visited
callees in a non-deterministic order. This makes it harder to write tests that
rely on the order of the output, and we also had crashes that were happening
only when we visited the callees in a certain order. To fix this, we consistently
put callees in a list as we discover them and use the previous hash map to keep
track of where a given callee is on the list so we can add call sites to it later.

* Source/WebGPU/WGSL/CallGraph.cpp:
(WGSL::CallGraphBuilder::visit):

Canonical link: https://commits.webkit.org/264932@main
